src/util/id.rs

```diff
@@ -1,33 +1,311 @@
+pub struct LocalId(NonZeroU64);
+
+impl LocalId {
+    /// Creates a local id for a raw value.
+    ///
+    /// The value must not be 0.
+    pub const fn from_raw(value: u64) -> Self {
+        if value == 0u64 {
+            panic!("Local id must not be 0");
+        }
+
+        unsafe { // Need const unwrap
+            Self(NonZeroU64::new_unchecked(value))
+        }
+    }
+
+    /// Creates a local id from a hash value. if the hash is 0 it will be set to 1.
+    pub const fn from_hash(mut hash: u64) -> Self {
+        if hash == 0u64 {
+            hash = 1u64;
+        }
+
+        unsafe { // Need const unwrap
+            Self(NonZeroU64::new_unchecked(hash))
+        }
+    }
+
+    pub const fn get_raw(&self) -> u64 {
+        self.0.get()
+    }
+}
+
+impl Debug for LocalId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("LocalId").field(&self.get_raw()).finish()
+    }
+}
+
+/// A universally unique identified.
+///
+/// A uuid is made up of a global id, local id pair.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct UUID {
+    pub global: GlobalId,
+    pub local: LocalId,
+}
+
+/// A utility struct providing a simple incrementing counter local id generator.
+///
+/// The generator will create its own global id. Local ids will be generated from a incrementing
+/// counter.
+///
+/// # Examples
+/// ```
+/// use rosella_rs::util::id::*;
+///
+/// // Create a new generator
+/// let generator = IncrementingGenerator::new();
+///
+/// // A new uuid
+/// let some_uuid = generator.next();
+///
+/// // The global id of the generator will be used for uuids
+/// assert_eq!(generator.get_global_id(), some_uuid.global);
+///
+/// // Some other uuid
+/// let other_uuid = generator.next();
+/// assert_ne!(some_uuid, other_uuid);
+/// ```
+pub struct IncrementingGenerator {
+    global: GlobalId,
+    next: std::sync::atomic::AtomicU64,
+}
+
+impl IncrementingGenerator {
+    /// Creates a new generator with a new global id and a local id starting at 0.
+    pub fn new() -> Self {
+        Self {
+            global: GlobalId::new(),
+            next: std::sync::atomic::AtomicU64::new(1),
+        }
+    }
+
+    /// Returns the global id of the generator.
+    pub fn get_global_id(&self) -> GlobalId {
+        self.global
+    }
+
+    /// Creates a new uuid
+    pub fn next(&self) -> Option<UUID> {
+        let local = self.next.fetch_add(1u64, std::sync::atomic::Ordering::Relaxed);
+
+        Some(UUID {
+            global: self.global,
+            local: LocalId::from_raw(local),
+        })
+    }
+}
+
+/// A UUID generated from a string.
+///
+/// NamedUUIDs use a predefined global id with the local id being calculated as the hash of a
+/// string. The name is stored along side the UUID for easy debugging or printing. The name is
+/// stored by Arc enabling fast Copying of the struct.
+#[derive(Clone, Debug)]
+pub struct NamedUUID {
+    name: Arc<String>,
```

Comment:
Wouldn't an Arc have more overhead than a String clone?

Reply:
A string clone needs to allocate new memory and copy the entire string.
A arc clone is incrementing a atomic int